File: C++/PalindromePermutation.cc
// Given a string, determine if a permutation of the string could form a palindrome.

// For example,
// "code" -> False, "aab" -> True, "carerac" -> True.

class Solution {
public:
    bool canPermutePalindrome(string s) {
        unordered_map<char, int> char_table;
        for (char c : s) {
            auto pair = char_table.emplace(c, 1);
            if (!pair.second) char_table[c]++;
        }
        bool single = s.size() % 2 == 0 ? true : false;
        for (auto p : char_table) {
            if (p.second % 2 != 0) {
                if (single) return false;
                else single = true;
            }
        }
        return true;
    }
};

File: Java/PalindromePermutation.java
// Given a string, determine if a permutation of the string could form a palindrome.

// For example,
// "code" -> False, "aab" -> True, "carerac" -> True.

public class PalindromePermutation {
    public boolean canPermutePalindrome(String s) {
        char[] sa = s.toCharArray();
        int[] c = new int[256];
        for (int i = 0; i < sa.length; i++) {
            c[sa[i]]++;
        }
        int count = 0;
        for (int i :c) {
            if (i % 2 == 1 && ++count > 1) {
                return false;
            }
        }
        return true;
    }
}
